overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,130 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
+	keep := func(name string, rev snap.Revision) {
+		snapToRevisions[name] = append(snapToRevisions[name], rev)
+	}
+
+	// keep revisions in snap's sequence
+	for snapName, snapst := range snapStates {
+		for _, seq := range snapst.Sequence {
+			keep(snapName, seq.Revision)
+		}
+	}
+
+	// keep revisions in refresh hints
+	for snapName, hint := range refreshHints {
+		keep(snapName, hint.Revision())
+	}
+
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, _, err := snapSetupAndState(t)
+			if err != nil {
+				return nil, err
+			}
+			keep(snapsup.InstanceName(), snapsup.Revision())
+		}
+	}
+
+	return snapToRevisions, nil
+}
+
+// snapRevisionsToKeep returns revisions that need to be kept for the given snap.
+//
+// It is the caller's responsibility to lock state before calling this function.
+func snapRevisionsToKeep(st *state.State, snapName string) ([]snap.Revision, error) {
+	snapToRevisions, err := revisionsToKeep(st)
+	if err != nil {
+		return nil, err
+	}
+	return snapToRevisions[snapName], nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
+	snapToRevisions, err := revisionsToKeep(st)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for snapName, revisions := range snapToRevisions {
+		for _, revision := range revisions {
+			mountFile := snap.MountFile(snapName, revision)
+			if osutil.FileExists(mountFile) {
+				keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirState(dirs.SnapBlobDir, "*.snap", keepContent)
+	return err
+}
+
+// cleanSnapDownloads removes downloads that are no longer needed for the given snap.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanSnapDownloads = func(st *state.State, snapName string) error {
+	revisions, err := snapRevisionsToKeep(st, snapName)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for _, revision := range revisions {
+		mountFile := snap.MountFile(snapName, revision)
+		if osutil.FileExists(mountFile) {
+			keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+		}
+	}
+
+	// XXX: could we need more digits?
+	const maxRevDigits = 10
+	var globs []string
+	for i := 1; i <= maxRevDigits; i++ {
+		digitsPat := strings.Repeat("[0-9]", i)
+		globs = append(globs, fmt.Sprintf("%s_%s.snap", snapName, digitsPat))
+		globs = append(globs, fmt.Sprintf("%s_x%s.snap", snapName, digitsPat))
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapBlobDir, globs, keepContent)
```

Comment:
doing a Glob on \<instance_Name>_*.snap and then filtering by what to keep and a regexp, and then doing os.Remove might be simpler again, also because this as again the compare with itself issue

Reply:
updated, thank you!